torch/utils/_mode_utils.py

```diff
@@ -0,0 +1,123 @@
+import functools
+from typing import Iterator
+from dataclasses import dataclass
+
+
+def _wrap_init(f, class_name, mode_type):
+    undef = object()
+
+    @functools.wraps(f)
+    def wrapped(self, *args, inner=undef, **kwargs):
+        if inner is undef:
+            raise TypeError(
+                f"missing inner keyword argument; instead of constructing a {class_name} directly, "
+                f"pass the constructor to push_{mode_type}_mode"
+            )
+        self.inner = inner
+        return f(self, *args, **kwargs)
+    return wrapped
+
+
+class ModeMeta(type):
+    def __new__(metacls, name, bases, dct, mode_type=None):
+        if mode_type not in ['torch_function', 'python']:
+            raise RuntimeError(f"only support torch_function or python modes, got mode_type of {mode_type}")
+        if '__init__' in dct:
+            dct['__init__'] = _wrap_init(dct['__init__'], metacls, mode_type)
+        return super().__new__(metacls, name, bases, dct)
+
+
+# in order to dedupe the logic between python mode and torch_function mode, this
+# is a container to hold all the differences between the modes. Then functions like
+# _enable_mode are able to use this container to call functions or get correctly
+# formatted names
+@dataclass
+class _ModeInfo:
+    mode_name: str
+    mode_class: type  # the class related to the mode that's allowed to be passed in
+    base_mode_class: type  # the base class of mode_class that dispatches to the original function
+    required_fn: str  # string version of the function required, either torch_function or torch_dispatch
+
+
+    def mode_class_name(self):
+        return self.mode_class.__name__
+
+    def get_mode(self):
+        """gets the current mode for this type of mode"""
+        raise NotImplementedError()
+
+    def set_mode(self, mode):
+        """
+        set mode to for this type of mode. Note that no checks are done on this, it's the unsafe
+        version where checks are assumed to have been already done by the helper function
+        """
+        raise NotImplementedError()
+
+
+# shared version of enable_torch_function/enable_python_mode in order to deduplicate the code.
+# The differences between the modes are captured by `mode_info` and then queried when they're
+# needed during the function's invocation
+def _enable_mode(mode, mode_info: _ModeInfo, *, replace=None, ignore_preexisting=False) -> Iterator[None]:
+    if not (
+        mode is None or
+        isinstance(mode, mode_info.mode_class) or
+        (isinstance(mode, type) and not issubclass(mode, mode_info.mode_class))
+    ):
+        raise ValueError(f'expected to get {mode_info.mode_class_name()}, Tensor-like class, '
```

Comment:
for dispatch, the test for Tensor-like can be more stringent, as only subclasses of torch.Tensor can be modes. Still wondering if I should have just eliminated this case though lol

Reply:
Do you mean the case where we allow passing in subclasses of Tensor rather than a Mode object? 

If so, my vote is to eliminate those cases: the way someone writes a Mode's `__torch_function__/__torch_dispatch__` is different from the way someone writes a Tensor subclass's `__torch_function__/__torch_dispatch__`. In particular, for Tensors, (1) there's an implicit assumption that at least one of the Tensors being dispatched on is the Tensor subclass and (2) the torch_function/torch_dispatch on a Tensor subclass is recommended to be classmethod (so, it cannot store metadata unlike a Mode object)

Reply:
Stacked this on #76562, naming should be consistent now!